Adapt Tarjan generic cycle detector for use in Crossgen2

trylek opened the pull request:
This change modifies Crossgen2 to use the generic cycle detector
originally implemented for NativeAOT to trim infinite generic
expansion as observed in the LanguageExt public nuget package
and tracked by the issue

https://github.com/dotnet/runtime/issues/66079

At this point I have only implemented a back-compat option to
opt out of the generic cycle detection in case it turns out to
cause problems for certain payloads. I haven't added the option
to set the cutoff level - in practice, LanguageExt compilation
still fails when the cutoff is set to anything higher than 0;
moreover for Crossgen2 skipping precompilation of certain methods
merely incurs slightly reduced performance, not runtime failures
as is the case with NativeAOT.

Thanks

Tomas

Fixes: https://github.com/dotnet/runtime/issues/66079

/cc @dotnet/crossgen-contrib

trylek committed:
Adapt Tarjan generic cycle detector for use in Crossgen2

This change modifies Crossgen2 to use the generic cycle detector
originally implemented for NativeAOT to trim infinite generic
expansion as observed in the LanguageExt public nuget package
and tracked by the issue

https://github.com/dotnet/runtime/issues/66079

At this point I have only implemented a back-compat option to
opt out of the generic cycle detection in case it turns out to
cause problems for certain payloads. I haven't added the option
to set the cutoff level - in practice, LanguageExt compilation
still fails when the cutoff is set to anything higher than 0;
moreover for Crossgen2 skipping precompilation of certain methods
merely incurs slightly reduced performance, not runtime failures
as is the case with NativeAOT.

Thanks

Tomas

Rebase against latest main

Temporary instrumentation to track dependency level

Improve dependency level instrumentation

Instrumentation for dumping dropped cycles

Add more places for cycle detection

More instrumentation

Log generic cycles for each module

MichalStrehovsky commented:
> LanguageExt compilation still fails when the cutoff is set to anything higher than 0

You might need to add a callout to the detection in more than one place. If anything higher than 0 doesn't work, I suspect the problem is that we don't have a callout in all the places that need it. I would at least try to look where the cycle is forming if we have a higher cutoff because that part of crossgen2 might be reachable if user code has a different kind of cycle.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:
```diff
@@ -584,6 +590,11 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
                             corInfoImpl.CompileMethod(methodCodeNodeNeedingCode, Logger);
                         }
                     }
+                    catch (Internal.TypeSystem.TypeSystemException.TypeLoadException)
+                    {
+                        if (Logger.IsVerbose)
+                            Logger.Writer.WriteLine($"Warning: Method `{method}` was not compiled because it participates in a generic cycle");
+                    }
```
I would just let it fall through to the `catch (TypeSystemException ex)` below. `TypeLoadException` can signify other kind of problem. It doesn't have to be a generic cycle.

trylek replied:
Good suggestion, thanks, fixed in the latest commit.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:
```diff
@@ -584,6 +590,11 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
                             corInfoImpl.CompileMethod(methodCodeNodeNeedingCode, Logger);
                         }
                     }
+                    catch (Internal.TypeSystem.TypeSystemException.TypeLoadException)
```
I agree with Michal here. We cannot catch `TypeLoadException` and assume it is associated with the `GenericCycleDetector`. You either need to wrap the calls to `GenericCycleDetector` in a try/catch and translate the exception to something well known, in order to print this out, or modify the `GenericCycleDetector` to produce a unique failure that can be identified.

trylek replied:
Removed special handling per Michal's PR feedback in the latest commit.

davidwrighton reviewed (changes requested):
We need cleaner error handling here.

mangod9 commented:
@trylek is this still being considered or should be closed?

trylek commented:
It should be ultimately fixed but we can close it for now, I'll reopen this once we're done with the experiments and start working on .NET 8 features.

trylek committed:
Mop-up fixes after rebase to most recent main

trylek committed:
Implement breadth cutoff; remove temporary instrumentation

trylek committed:
Remove remaining instrumentation and consolidate command-line args

trylek commented:
I believe I have finally managed to make this work for the repro case described in https://github.com/dotnet/runtime/issues/66079. The main trick is that the depth cutoff heuristic doesn't work well in the case of LanguageExt.Core - the problem with this assembly is that many generic types have lots of type parameters and what basically happens is a form of exponential explosion not necessarily recursively repeating a given generic type within a single type parameter. To cater for this I have introduced a second heuristic - breadth cutoff - that corresponds to the total number of occurrences of types marked as poisoned w.r.t. potentially forming cycles within a generic type that cause the dependency to be identified as potentially cycle-forming. With the default value of 4 for the depth cutoff I have measured sizes of the compiled LanguageExt.Core assembly (Windows x64 release) variating the breadth cutoff:

| Breadth cutoff | LanguageExt.Core Size |
| ---------------- | ------------------------- |
| 1                      | 32 MB |
| 2                      | 49 MB |
| 3                      | 120 MB |
| 4                      | 592 MB |

In light of these findings I set the default breadth cutoff to 2. After discussing the matter with Manish I have come to the conclusion that for now it's probably better to make this just an opt-in especially in light of the fact that the compilation takes longer due to the need to build the module tables of poisoned types. If at some point we decide to make this the default (possibly with an opt-out switch) there are several places where we might be able to optimize the algorithm - e.g. by parallelizing the initial poisoned types generation, by caching the results of IsDeepPossiblyCyclicInstantiation, by improving the quadratic algorithm for checking the depth cutoff and the like. I'm still working on validating the change in the lab but for now I believe that it should be feature complete, so to say.

trylek commented:
One other aspect worth mentioning is that at this point the change is slightly hacky by manually bringing in bits of NativeAOT logic pertinent to the generic cycle detection along with bits of logging logic. I think that it would be useful to unify this between Crossgen2 and NativeAOT by either putting the relevant code in one of the existing common projects or by creating a new one shared between the two compilers.

trylek committed:
Fix incorrect rebase

trylek committed:
One more rebase fix

trylek committed:
Fix NativeAOT compiler variant of the generic cutoff point

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:
```diff
@@ -135,7 +135,7 @@ internal sealed class ILCompilerRootCommand : RootCommand
             new(new[] { "--directpinvokelist" }, Array.Empty<string>, "File with list of PInvokes to call directly");
         public Option<int> MaxGenericCycle { get; } =
-            new(new[] { "--maxgenericcycle" }, () => CompilerTypeSystemContext.DefaultGenericCycleCutoffPoint, "Max depth of generic cycle");
+            new(new[] { "--maxgenericcycle" }, () => CompilerTypeSystemContext.DefaultGenericCycleDepthCutoff, "Max depth of generic cycle");
```
Could you add a new argument to control the other cutoff? If I'm reading it right, the new cutoff also applies to NativeAOT and we should have a way to control it if it misbehaves.

I don't have a good mental model of how the new cutoff will work in practice - maybe we should also have a `-1` value that means "don't check for this".

trylek replied:
Added in 8th commit.

MichalStrehovsky commented:
Is it possible to write a test where crossgen would previously fail and now works? We have some NativeAOT specific testing in smoketests for generic recursions that were killing the NAOT compiler, mostly extracted from the real world patterns.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/ILCompiler.ReadyToRun.csproj:
```diff
@@ -148,6 +152,15 @@
+    <Compile Include="..\ILCompiler.Compiler\Compiler\LazyGenerics\GraphBuilder.ForEach.cs" Link="Compiler\GenericCycleDetection\GraphBuilder.ForEach.cs" />
+    <Compile Include="..\ILCompiler.Compiler\Compiler\LazyGenerics\GraphBuilder.MethodCall.cs" Link="Compiler\GenericCycleDetection\GraphBuilder.MethodCall.cs" />
+    <Compile Include="..\ILCompiler.Compiler\Compiler\LazyGenerics\ModuleCycleInfo.cs" Link="Compiler\GenericCycleDetection\ModuleCycleInfo.cs" />
+    <Compile Include="..\ILCompiler.Compiler\Compiler\Logging\MessageOrigin.cs" Link="Compiler\Logging\MessageOrigin.cs" />
+    <Compile Include="..\ILCompiler.Compiler\Compiler\Logging\MessageContainer.cs" Link="Compiler\Logging\MessageContainer.cs" />
```
The reason to not share these with crossgen was that ilc has the entire warning subsystem in it - suppressions, attributes, logging MSBuild friendly format, command line options for this, ...
Crossgen doesn't have any of that and I didn't see a reason to add that, so I tried to keep the two logging systems as separate as possible.

I don't think crossgen needs/wants all of those subsystems yet. So we can either ifdef this (and there will be no diagnostics printed out from crossgen with this algorithm), or we refactor the logging somewhat to make it possible to be used even without all the subsystems. I didn't spent time to look into what that would take though.

So I think we should decide what is the behavior we want from crossgen logging:
* Do we expect it to produce real warnings (with warning codes and so on)?
* If no, then I would ifdef this out.
* If we think it should produce real warnings, then we should probably create a separate issue which will implement at least some of the warning subsystems into crossgen (command line suppressions, warning codes, MSBuild friendly formatting).

trylek replied:
I have added conditional compilation clauses that drop the logging of generic cycle warnings in Crossgen2. After all, in comparison with NativeAOT this is less detrimental in Crossgen2 where we can always resort to runtime JIT. As the last bit of feedback I'm aware of I'm going to work on implementing a regression test for the issue.

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/TypeSystemContext.cs:
```diff
@@ -770,6 +770,18 @@ internal TypeFlags ComputeTypeFlags(TypeDesc type, TypeFlags flags, TypeFlags ma
+        /// <param name="callee">The new dependency to be added to the graph</param>
+        /// <param name="name">Textual name describing the dependency</param>
+        /// <exception cref="NotSupportedException"></exception>
+        public virtual void DetectGenericCycles(TypeSystemEntity caller, TypeSystemEntity callee, string name)
```
Thanks for the suggestion. I have ultimately ended up moving the method to <code>NodeFactory</code> as that is readily accessible from most of about 10 places in the code that call the method. There was only one place where <code>NodeFactory</code> wasn't easily accessible, <code>ReadyToRunModuleCompilationGroupBase</code> so I moved the generic cycle check from there to the method <code>CanInline</code> in <code>ReadyToRunCodegenCompilation</code>. I'll publish this as part of the other conservative changes in 8th commit, then I'll work on building the regression test - we probably cannot directly use the LanguageExt.Core library as that's an external package; I think I have an idea how to build the test based on watching the generic expansions in LanguageExt.Core.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilerContext.cs:
```diff
@@ -109,6 +137,11 @@ public void SetCompilationGroup(ReadyToRunCompilationModuleGroupBase compilation
             _r2rFieldLayoutAlgorithm.SetCompilationGroup(compilationModuleGroup);
         }
 
+        public override void DetectGenericCycles(TypeSystemEntity caller, TypeSystemEntity callee, string name)
```
I have removed them in the 8th commit, I doubt they would be really actionable for normal developers even if we emitted them in some detailed logging.

trylek commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CompilerTypeSystemContext.Aot.cs:
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+    // Licensed to the .NET Foundation under one or more agreements.
```
Sorry about the typo, fixed in 8th commit.

trylek committed:
Address Michal's local PR feedback (simple functionality shuffles)

I'll address the code reorg (GenericCycleDetection) and regression
test in subsequent commits.

Thanks

Tomas

trylek committed:
Revert change to pre-existing command-line option naming to reduce churn

trylek committed:
Move cycle detector under GenericCycleDetection per Michal's PR feedback

trylek committed:
Delete no longer needed logging changes

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/ILCompiler.DependencyAnalysisFramework.csproj:
```diff
@@ -8,6 +8,7 @@
     <Platforms>x64;x86</Platforms>
     <PlatformTarget>AnyCPU</PlatformTarget>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <DefineConstants>READYTORUN;$(DefineConstants)</DefineConstants>
```
This project is shared. This doesn't look right

trylek committed:
Revert incorrect addition of READYTORUN define from DependencyAnalysisFramework

trylek committed:
Add generic depth cutoff test

So I have a trivial depth cutoff test working; without enabling
generic cycle detection Crossgen2 crashes after about an hour on my
box with the Arithmetic Overflow as described in the issue #66079.
As next step I'll work on the equivalent breadth test, that is
somewhat more tricky.

Thanks

Tomas

trylek committed:
Adjust DepthTest so that it only uses the depth cutoff

trylek committed:
Implement unit tests for generic cycle detection

trylek committed:
Fix generic breadth test

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:
```diff
@@ -106,6 +106,15 @@ public bool CanInline(MethodDesc caller, MethodDesc callee)
+            {
+                _nodeFactory.DetectGenericCycles(caller, callee);
+            }
+            catch (TypeSystemException)
```
Wouldn't this exception be handled in JitInterface?

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/MethodWithGCInfo.cs:
```diff
@@ -271,7 +271,19 @@ protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFact
+                    dependencyList.Add(node, "classMustBeLoadedBeforeCodeIsRun");
+
+                }
+                catch (Exception)
```
```suggestion
                catch (TypeSystemException)
```

trylek replied:
Good point, fixed in 17th commit.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/MethodWithGCInfo.cs:
```diff
@@ -271,7 +271,19 @@ protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFact
 
             foreach (ISymbolNode node in _fixups)
```
On the NativeAOT side we do this check as we compile the method - it ends up invalidating the whole body. Dropping the fixup here makes me uneasy but I don't know that much about R2R.

trylek replied:
I don't think we're actually dropping the fixup, we're just not using it for further dependency tree expansion i.o.w. methods only referred to via such fixup would not be precompiled and would require runtime JIT; please note this only pertains to the "pathologic" case of methods potentially causing generic cycles.

davidwrighton replied:
This is dropping the dependencies of the fixup. I'm not very comfortable with that, and agree with @MichalStrehovsky here. Instead, could we detect the cycle when adding the fixup to _fixups, and if we do so, mark the method with whatever details we do, when then entire method can't be compiled.

MichalStrehovsky reviewed (approved):
The changes to shared/ILC specific files look good to me. I didn't pay close attention to the crossgen2 specific parts as I can't speak to how it's hooked up. Would be good to get a second pair of eyes on that.

trylek committed:
Address Michal's PR feedback, fix one unit test to avoid lab timeouts

trylek committed:
Make Depth3Test unsupported on 32-bit platforms where it OOMS Crossgen2

trylek committed:
Fix typo in MSBuild condition

davidwrighton reviewed (changes requested):
I'd like the handling of the fixups for compiled methods to be tweaked a bit. Also, do we know what impact this has on the performance of crossgen?

trylek commented:
Thanks David for your feedback, I'll modify the fixup management based on your suggestion. The impact on Crossgen2 performance is quite detrimental, that's why it's just an opt-in for now, if we decided to turn it on by default, we would at the very least need to improve the perf quite a bit. For the LanguageExt assembly mentioned in the original issue, just the module analysis takes something like a minute on my devbox.

MichalStrehovsky commented:
> just the module analysis takes something like a minute on my devbox

Is that a debug build? From what I saw when enabling this, it usually takes single digits milliseconds to analyze an assembly.

trylek committed:
Reorganize handling of method fixups per David's PR feedback

trylek commented:
Thanks David and Michal for your feedback. I have updated the PR along David's suggestion. For Michal's comment, I have instrumented <code>CycleInfoHashtable</code> to show time spent in the analysis and this is the result for the <code>LanguageExt.Core</code> assembly from the original issue repro:

<pre>
Cycle analysis for LanguageExt.Core took 360382 msecs
Cycle analysis for System.Private.CoreLib took 45 msecs
Cycle analysis for LanguageExt.Core took 363084 msecs
Cycle analysis for LanguageExt.Core took 363232 msecs
Cycle analysis for LanguageExt.Core took 363388 msecs
Cycle analysis for LanguageExt.Core took 364608 msecs
Cycle analysis for LanguageExt.Core took 365848 msecs
Cycle analysis for LanguageExt.Core took 367601 msecs
Cycle analysis for LanguageExt.Core took 368732 msecs
</pre>

Apparently one problem is that the internal Crossgen2 parallelism triggers initial analysis of the assembly on all the threads by virtue of the lock-free hashtable; I'm not sure if some form of locking would improve this. This is kind of the slowest of my devboxes but I doubt its <code>Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz</code> is a thousand times slower than Michal's box. I think one problem could be that the LanguageExt.Core assembly is a real large one, about 10 MB or on par with CoreLib, just much more complex in terms of heavy use of generic structs everywhere; please feel free to try it out in NativeAOT mode, I don't see how the Tarjan cycle detector could be so much faster there considering it basically just uses the ECMA / typesystem code that is shared between the two compilers.

trylek commented:
I have experimentally added a somewhat absurd lock around the lock-free hashtable and it indeed made things go faster in general:

<pre>
Cycle analysis for LanguageExt.Core took 96969 msecs
Cycle analysis for System.Private.CoreLib took 32 msecs
Emitting R2R PE file: D:\git\AndreSteenveld.CrossgenLanguageExt\obj\Release\net6.0\win-x64\R2R\LanguageExt.Core.dll
</pre>

In light of this fact I think it might make sense to consider dropping the parallelized hashtable population in this particular case but technically such change would be somewhat orthogonal to my primary effort and I would feel comfortable making it in a separate change if we agree it's the right way to go. Also my somewhat ham-fisted approach globally locks the hashtable whereas a more subtle fix could just lock its population for a particular module, I think that two different modules can be happily scanned in parallel.

trylek commented:
(Just to clarify regarding Michal's above question, these are measurements on Windows.x64.Release build.)

MichalStrehovsky commented:
Oh, if it's a 10 MB assembly, I'm not too concerned. This is absolutely huge by any standards. CoreLib is a good example of a "normal" big assembly and that's 45 ms, which is the expected ballpark.

The locking may or may not benefit things. Normal sized assemblies take single digits ms and that's comparable with jitting a couple methods. It would only help if we somehow figure out things will take longer (maybe from size of typespec table?)

trylek commented:
Thanks Michal for your additional feedback. Looking at the compilation in the debugger, I see that <code>LanguageExt.Core</code> has <code> | NumberOfRows | 0x000090ed | </code> (about 37K) in the <code>TypeSpec</code> table; in contrast, SPC has just <code> | NumberOfRows | 0x00000636 | </code> (1.5K or so) in the same table. The interesting question is how to actually utilize this number in the compilation - we could possibly treat it as an indicator that we need to run the generic cycle analysis, to avoid parallelization in its initial scanning and / or for additional tweaks in the compilation pipeline. We'd also need to establish and measure additional heuristics like - is this supposed to be compared with a hardcoded constant, or somehow weighted as a percentage of the number of TypeDesc / MethodDesc table entries or something yet completely different.
